torch.compile() fails at TransformerEngineBaseModule.prepare_forward()

`torch.compile(model).forward(x)` fails if `TransformerEngineBaseModule.prepare_forward()` is used as a context manager during the call to `model.forward()`. 

Here's a simple modification of the example code from the README that reproduces this error:

```
import torch
import transformer_engine.pytorch as te
from transformer_engine.common import recipe

# Set dimensions.
in_features = 768
out_features = 3072
hidden_size = 2048

# Initialize model and inputs.
model = te.Linear(in_features, out_features, bias=True)
inp = torch.randn(hidden_size, in_features, device="cuda")

# Create an FP8 recipe. Note: All input args are optional.
fp8_recipe = recipe.DelayedScaling(margin=0, interval=1, fp8_format=recipe.Format.E4M3)

# Enable autocasting for the forward pass
model = torch.compile(model)
with te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe):
    out = model(inp)

loss = out.sum()
loss.backward()
```
Only one line is added to the original: `model = torch.compile(model)`. 

Running this in version 23.06-py3 of [the Pytorch container from the NGC catalogue](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch) using an RTX 4090 produces the following error:

```
root@048d175db259:/workspace/te_example# python te_example.py 
Traceback (most recent call last):
  File "/workspace/te_example/te_example.py", line 20, in <module>
    out = model(inp)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 280, in _fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/transformer_engine/pytorch/module/linear.py", line 706, in forward
    with self.prepare_forward(inp, is_first_microbatch) as inp:
  File "/usr/local/lib/python3.10/dist-packages/transformer_engine/pytorch/module/linear.py", line 706, in <resume in forward>
    with self.prepare_forward(inp, is_first_microbatch) as inp:
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.10/dist-packages/transformer_engine/pytorch/module/base.py", line 605, in prepare_forward
    add_amax_to_global_buffer(self.fp8_meta, forward=True)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 433, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/convert_frame.py", line 501, in _convert_frame
    result = inner_convert(frame, cache_size, hooks, frame_state)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/convert_frame.py", line 122, in _fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/convert_frame.py", line 337, in _convert_frame_assert
    return _compile(
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/utils.py", line 177, in time_wrapper
    r = func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/convert_frame.py", line 407, in _compile
    out_code = transform_code_object(code, transform)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/convert_frame.py", line 392, in transform
    tracer.run()
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/symbolic_convert.py", line 2006, in run
    super().run()
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/symbolic_convert.py", line 703, in run
    and self.step()
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/symbolic_convert.py", line 663, in step
    getattr(self, inst.opname)(inst)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/symbolic_convert.py", line 1171, in LOAD_ATTR
    result = BuiltinVariable(getattr).call_function(
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/variables/builtin.py", line 584, in call_function
    result = handler(tx, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/variables/builtin.py", line 1056, in call_getattr
    obj.var_getattr(tx, name).clone(source=source).add_options(options)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/variables/user_defined.py", line 392, in var_getattr
    return variables.UserMethodVariable(
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/variables/base.py", line 27, in __call__
    obj = type.__call__(cls, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/variables/functions.py", line 276, in __init__
    super().__init__(fn=fn, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/variables/functions.py", line 124, in __init__
    assert isinstance(
AssertionError: expected FunctionType found instancemethod <instancemethod  at 0x7f9e1cebe7a0>

from user code:
   File "/usr/local/lib/python3.10/dist-packages/transformer_engine/pytorch/fp8.py", line 125, in add_amax_to_global_buffer
    _global_fp8_buffer[buffer_key] = [fp8_meta[fp8_meta_tensor_key].amax_history[0]]
```

I *think* the problem is that `TransformerEngineBaseModule.prepare_forward` is a method that is used as a context manager, and not a function. The [code in pytorch](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/functions.py#L143) that checks this is
```
 assert isinstance(
            fn, (types.FunctionType, torch.jit.ScriptFunction)
        ), f"expected FunctionType found {typestr(fn)} {fn}"
```
and a method is `types.MethodType`, not `types.FunctionType`.

I'm surprised no one else ran into this already? Did I make a mistake? 

It seems pretty important because it means that it's not possible to compile any models that use modules from transformer_engine. I initially ran into it when trying to see if transformer_engine could speed up [Karpathy's nanoGPT](https://github.com/karpathy/nanoGPT).

--- a/qa/L0_unittest/test.sh
+++ b/qa/L0_unittest/test.sh
@@ -10,3 +10,4 @@ pip install pytest==6.2.5 onnxruntime==1.13.1
 pytest -v -s $TE_PATH/tests/pytorch/test_sanity.py
 PYTORCH_JIT=0 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0 pytest -v -s $TE_PATH/tests/pytorch/test_numerics.py
 NVTE_TORCH_COMPILE=0 pytest -v -s $TE_PATH/tests/pytorch/test_onnx_export.py
+pytest -v -s $TE_PATH/tests/pytorch/test_jit.py

--- a/tests/pytorch/test_jit.py
+++ b/tests/pytorch/test_jit.py
@@ -12,6 +12,8 @@ import transformer_engine.pytorch as te
 # Model names for test_torch_dynamo
 _model_names = ["Linear", "LayerNorm", "LayerNormLinear", "LayerNormMLP"]
 
+
+@pytest.mark.skipif(torch.__version__ < "2", reason="torch.compile not available")
 @pytest.mark.parametrize("model_name", _model_names)
 def test_torch_dynamo(model_name: str):
     """Test compatibility with Torch Dynamo

--- a/tests/pytorch/test_jit.py
+++ b/tests/pytorch/test_jit.py
@@ -1,0 +1,62 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+from typing import Tuple
+
+import pytest
+import torch
+
+import transformer_engine.pytorch as te
+
+# Model names for test_torch_dynamo
+_model_names = ["Linear", "LayerNorm", "LayerNormLinear", "LayerNormMLP"]
+
+@pytest.mark.parametrize("model_name", _model_names)
+def test_torch_dynamo(model_name: str):
+    """Test compatibility with Torch Dynamo
+
+    Construct model, optimize with Torch Dynamo, and perform a single
+    forward and backward pass.
+
+    """
+
+    # Helper function to construct tensor with default options
+    def make_tensor(
+            dims: Tuple[int],
+            dtype: torch.dtype = torch.float32,
+            device: torch.device = "cuda",
+            requires_grad: bool = True,
+            **kwargs,
+    ):
+        return torch.zeros(
+            dims,
+            dtype=dtype,
+            device=device,
+            requires_grad=requires_grad,
+            **kwargs,
+        )
+
+    # Construct model and input tensors
+    model = None
+    inputs = []
+    if model_name == "Linear":
+        model = te.Linear(16, 16)
+        inputs = [make_tensor([16,16])]
+    elif model_name == "LayerNorm":
+        model = te.LayerNorm(16)
+        inputs = [make_tensor([16,16])]
+    elif model_name == "LayerNormLinear":
+        model = te.LayerNormLinear(16,16)
+        inputs = [make_tensor([16,16])]
+    elif model_name == "LayerNormMLP":
+        model = te.LayerNormMLP(16,16)
+        inputs = [make_tensor([16,16])]
+    assert model is not None, f"could not construct {model_name}"
+
+    # Optimize model with TorchDynamo
+    torch.compile(model)
+
+    # Forward and backward pass
+    out = model(*inputs)
+    out.backward(torch.zeros_like(out))

--- a/transformer_engine/pytorch/jit.py
+++ b/transformer_engine/pytorch/jit.py
@@ -12,6 +12,13 @@ jit_fuser = torch.jit.script
 if torch.__version__ >= "2" and bool(int(os.getenv("NVTE_TORCH_COMPILE", "1"))):
     jit_fuser = torch.compile
 
+# Decorator to disable Torch Dynamo
+# See: https://github.com/NVIDIA/TransformerEngine/issues/308
+no_torch_dynamo = lambda func: func
+if torch.__version__ >= "2":
+    import torch._dynamo
+    no_torch_dynamo = torch._dynamo.disable
+
 
 def set_jit_fusion_options() -> None:
     """Set PyTorch JIT layer fusion options."""

--- a/transformer_engine/pytorch/module/layernorm.py
+++ b/transformer_engine/pytorch/module/layernorm.py
@@ -12,6 +12,8 @@ from torch.nn import init
 
 import transformer_engine_extensions as tex
 
+from ..jit import no_torch_dynamo
+
 
 __all__ = ["LayerNorm"]
 
@@ -154,6 +156,7 @@ class LayerNorm(torch.nn.Module):
         init.zeros_(self.bias)
 
 
+    @no_torch_dynamo
     def forward(self, inp: torch.Tensor) -> torch.Tensor:
         """LayerNorm FWD"""
         # Maintain backward compatibility.

--- a/transformer_engine/pytorch/module/layernorm_linear.py
+++ b/transformer_engine/pytorch/module/layernorm_linear.py
@@ -48,6 +48,7 @@ from ..cpp_extensions import (
     cast_from_fp8,
 )
 from ..constants import GemmParallelModes, dist_group_type, TE_DType
+from ..jit import no_torch_dynamo
 
 
 __all__ = ["LayerNormLinear"]
@@ -820,6 +821,7 @@ class LayerNormLinear(TransformerEngineBaseModule):
 
         return fp8_weight_tensors
 
+    @no_torch_dynamo
     def forward(
         self,
         inp: torch.Tensor,

--- a/transformer_engine/pytorch/module/layernorm_mlp.py
+++ b/transformer_engine/pytorch/module/layernorm_mlp.py
@@ -44,6 +44,7 @@ from ..distributed import (
 from .. import cpp_extensions as tex
 
 from ..constants import dist_group_type, TE_DType
+from ..jit import no_torch_dynamo
 
 
 __all__ = ["LayerNormMLP"]
@@ -1144,6 +1145,7 @@ class LayerNormMLP(TransformerEngineBaseModule):
 
         return fp8_weight_tensors
 
+    @no_torch_dynamo
     def forward(
         self, inp: torch.Tensor, is_first_microbatch: Optional[bool] = None
     ) -> Union[torch.Tensor, Tuple[torch.Tensor, ...]]:

--- a/transformer_engine/pytorch/module/linear.py
+++ b/transformer_engine/pytorch/module/linear.py
@@ -42,6 +42,7 @@ from ..cpp_extensions import (
     cast_to_fp8,
 )
 from ..constants import GemmParallelModes, dist_group_type
+from ..jit import no_torch_dynamo
 
 
 __all__ = ["Linear"]
@@ -667,6 +668,7 @@ class Linear(TransformerEngineBaseModule):
 
         return fp8_weight_tensors
 
+    @no_torch_dynamo
     def forward(
         self,
         inp: torch.Tensor,

